src/main/java/org/cru/cas/client/integration/CopyCasAttributesToHeadersFilter.java

```diff
@@ -0,0 +1,251 @@
+        return headersToAttributeNames.entrySet()
+            .stream()
+            .collect(Collectors.toMap(
+                Entry::getValue,
+                entry -> Collections.singleton(entry.getKey()),
+                this::mergeSets
+            ));
+    }
+
+    private Set<String> mergeSets(Set<String> set1, Set<String> set2) {
+        return Stream.of(set1, set2)
+            .flatMap(Collection::stream)
+            .collect(Collectors.toSet());
+    }
+
+    @Override
+    public void doFilter(
+        final ServletRequest servletRequest,
+        final ServletResponse servletResponse,
+        final FilterChain filterChain
+    ) throws IOException, ServletException {
+
+        ServletRequest possiblyWrappedRequest = wrapServletRequestIfPossible(servletRequest);
+        filterChain.doFilter(possiblyWrappedRequest, servletResponse);
+    }
+
+    private ServletRequest wrapServletRequestIfPossible(ServletRequest servletRequest) {
+        if (servletRequest instanceof HttpServletRequest) {
+            HttpServletRequest httpServletRequest = (HttpServletRequest) servletRequest;
+            return new Wrapper(httpServletRequest);
+        } else {
+            return servletRequest;
+        }
+    }
+
+    private class Wrapper extends HttpServletRequestWrapper {
+
+        public Wrapper(HttpServletRequest request) {
+            super(request);
+        }
+
+        @Override
+        public Enumeration<String> getHeaderNames() {
+            Enumeration<String> originalHeaderNames = super.getHeaderNames();
+            if (originalHeaderNames == null) {
+                // container does not allow the header names to be enumerated
+                return null;
+            }
+            Set<String> allHeaderNames = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+            allHeaderNames.addAll(Collections.list(originalHeaderNames));
+            allHeaderNames.removeIf(name -> attributeMapping.headerIsMapped(name));
+            allHeaderNames.addAll(getAttributeHeaderNames());
+            return Collections.enumeration(allHeaderNames);
+        }
+
+        private Set<String> getAttributeHeaderNames() {
+            return getAttributes().entrySet()
+                .stream()
+                .filter(entry -> entry.getValue() != null)
+                .map(entry -> attributeMapping.getHeaderNamesForAttribute(entry.getKey()))
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
+        }
+
+        @Override
+        public Enumeration<String> getHeaders(String name) {
+            if (attributeMapping.headerIsMapped(name)) {
+                return Collections.enumeration(getHeaderValues(name));
+            } else {
+                return super.getHeaders(name);
+            }
+        }
+
+        @Override
+        public String getHeader(String name) {
+            if (attributeMapping.headerIsMapped(name)) {
+                return getFirstOrElseNull(getHeaderValues(name));
```

Comment:
Can you use a stream here instead? Something like `getHeadersValues(name).stream.findFirst().orElse(null);`

Reply:
Yeah, great suggestion.